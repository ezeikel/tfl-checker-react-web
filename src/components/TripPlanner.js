import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { useHistory, useLocation } from "react-router-dom";
import styled from "styled-components";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  fetchSuggestion,
  setFromCoords,
  setToCoords,
  setFromAddress,
  setToAddress,
  clearSuggestions,
} from "../redux/actions";
import GooglePlacesInput from "./GooglePlacesInput";
import TripSummaries from "./TripSummaries";
import { rotate } from "../GlobalStyle";

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

const JourneyInput = styled.section`
  display: flex;
  flex-direction: column;
  padding: 32px;
  background-color: var(--color-white);
  border-radius: var(--border-radius);
  max-width: 600px;

  hr {
    background-color: #aeb8c3;
    height: 1px;
    margin: 16px 0;
  }
`;

const Heading = styled.span`
  display: flex;
  color: var(--color-black);
  margin-bottom: 32px;
  font-size: 2.4rem;
  font-weight: bold;
`;

const FormWrapper = styled.div`
  background-color: var(--color-input-background);
  padding: 16px;
  margin-bottom: 32px;
  display: flex;
  border-radius: 4px;
  form {
    flex: 1 0 auto;
  }
`;

const InputWrapper = styled.div`
  position: relative;
  display: flex;
  flex-direction: column;
  label {
    font-size: 1.4rem;
    margin-bottom: 8px;
    color: #aeb8c3;
  }
`;

const StartEnd = styled.span`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-right: 16px;
  > svg {
    flex: 0 0 auto;
  }
`;

const VerticalLine = styled.span`
  flex: 1 0 auto;
  background-color: #aeb8c3;
  width: 1px;
`;

const Leaving = styled.span`
  display: flex;
  align-items: center;
  font-size: 1.6rem;
  margin-bottom: 32px;
  svg {
    width: 10px;
    height: 18px;
    margin-left: 8px;
  }
`;

const Button = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: var(--color-green);
  color: var(--color-white);
  font-size: var(--font-size-bravo);
  font-weight: bold;
  padding: 16px 0;
  border-radius: var(--border-radius);
  svg {
    margin-left: 16px;
    animation: ${rotate} 1.2s linear infinite;
  }
`;

const useQuery = () => new URLSearchParams(useLocation().search);

const TripPlanner = ({
  fromCoordinates,
  toCoordinates,
  fromAddress,
  toAddress,
  loading,
  results,
  onFetchSuggestions,
  onSetFromCoords,
  onSetToCoords,
  onSetFromAddress,
  onSetToAddress,
  onClearSuggestions,
}) => {
  const history = useHistory();
  const query = useQuery();

  useEffect(() => {
    if (
      fromAddress === "" &&
      toAddress === "" &&
      query.get("toCoordinates") &&
      query.get("fromCoordinates")
    ) {
      const queryParamsFromCoords = {
        lat: parseFloat(query.get("fromCoordinates").split(",")[0]),
        lng: parseFloat(query.get("fromCoordinates").split(",")[1]),
      };
      const queryParamsToCoords = {
        lat: parseFloat(query.get("toCoordinates").split(",")[0]),
        lng: parseFloat(query.get("toCoordinates").split(",")[1]),
      };
      const queryParamsFromAddress = query.get("fromAddress");
      const queryParamsToAddress = query.get("toAddress");

      onFetchSuggestions(queryParamsFromCoords, queryParamsToCoords);
      onSetFromCoords(queryParamsFromCoords);
      onSetToCoords(queryParamsToCoords);
      onSetFromAddress(queryParamsFromAddress);
      onSetToAddress(queryParamsToAddress);
    }
  }, [
    fromAddress,
    onFetchSuggestions,
    onSetFromAddress,
    onSetFromCoords,
    onSetToAddress,
    onSetToCoords,
    query,
    toAddress,
  ]);

  useEffect(() => {
    return () => {
      onClearSuggestions();
    };
  }, [onClearSuggestions]);

  const handleSubmit = () => {
    history.push(
      `/trip-planner?fromCoordinates=${fromCoordinates.lat},${fromCoordinates.lng}&toCoordinates=${toCoordinates.lat},${toCoordinates.lng}&fromAddress=${fromAddress}&toAddress=${toAddress}`,
    );
    onFetchSuggestions(fromCoordinates, toCoordinates);
  };

  return (
    <Wrapper>
      <JourneyInput>
        <Heading>Let&apos;s take a trip!</Heading>
        <FormWrapper>
          <StartEnd>
            <FontAwesomeIcon
              icon={["fad", "circle"]}
              color="var(--color-green)"
              size="2x"
            />
            <VerticalLine />
            <FontAwesomeIcon
              icon={["fad", "circle"]}
              color="var(--color-background)"
              size="2x"
            />
          </StartEnd>
          <form>
            <InputWrapper>
              {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
              <label htmlFor="fromAddress">From</label>{" "}
              <GooglePlacesInput
                inputId="fromAddress"
                setLocation={onSetFromCoords}
                address={fromAddress}
                setAddress={onSetFromAddress}
                placeholder="Where are you coming from?"
              />
            </InputWrapper>
            <hr />
            <InputWrapper>
              {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
              <label htmlFor="toAddress">To</label>
              <GooglePlacesInput
                inputId="toAddress"
                setLocation={onSetToCoords}
                address={toAddress}
                setAddress={onSetToAddress}
                placeholder="Where are you going to?"
              />
            </InputWrapper>
          </form>
        </FormWrapper>
        <Leaving>
          Leaving now
          <FontAwesomeIcon
            icon={["fad", "chevron-right"]}
            color="var(--color-background)"
            size="2x"
          />
        </Leaving>
        <Button onClick={handleSubmit}>
          {`Search${loading ? "ing" : ""}`}
          {loading && (
            <FontAwesomeIcon
              icon={["fad", "spinner-third"]}
              color="var(--color-white)"
              size="lg"
            />
          )}
        </Button>
      </JourneyInput>
      {// TODO: Disambiguations cause results to be null. Add a way to get exact location instead
      results && results.length ? <TripSummaries journeys={results} /> : null}
    </Wrapper>
  );
};

const mapStateToProps = ({ journey, suggestion }) => ({
  fromCoordinates: journey.from.coordinates,
  fromAddress: journey.from.address,
  toCoordinates: journey.to.coordinates,
  toAddress: journey.to.address,
  results: suggestion.results,
  loading: suggestion.loading,
});

const mapDispatchToProps = dispatch => ({
  onFetchSuggestions: (fromCoordinates, toCoordinates) =>
    dispatch(fetchSuggestion(fromCoordinates, toCoordinates)),
  onSetFromCoords: value => dispatch(setFromCoords(value)),
  onSetToCoords: value => dispatch(setToCoords(value)),
  onSetFromAddress: value => dispatch(setFromAddress(value)),
  onSetToAddress: value => dispatch(setToAddress(value)),
  onClearSuggestions: () => dispatch(clearSuggestions()),
});

TripPlanner.defaultProps = {
  fromCoordinates: {},
  toCoordinates: {},
  fromAddress: "",
  toAddress: "",
  results: [],
  loading: false,
};

TripPlanner.propTypes = {
  // TODO: Make sure lat/lng are one type: number
  fromCoordinates: PropTypes.shape({
    lat: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    lng: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  }),
  // TODO: Make sure lat/lng are one type: number
  toCoordinates: PropTypes.shape({
    lat: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    lng: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  }),
  fromAddress: PropTypes.string,
  toAddress: PropTypes.string,
  results: PropTypes.arrayOf(
    PropTypes.shape({
      $type: PropTypes.string,
      startDateTime: PropTypes.string,
      duration: PropTypes.number,
      arrivalTime: PropTypes.string,
    }),
  ),
  loading: PropTypes.bool,
  onFetchSuggestions: PropTypes.func.isRequired,
  onSetFromCoords: PropTypes.func.isRequired,
  onSetToCoords: PropTypes.func.isRequired,
  onSetFromAddress: PropTypes.func.isRequired,
  onSetToAddress: PropTypes.func.isRequired,
  onClearSuggestions: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(TripPlanner);
