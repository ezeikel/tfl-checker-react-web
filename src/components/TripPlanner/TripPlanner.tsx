import { useEffect } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useQuery } from "@tanstack/react-query";
import GooglePlacesInput from "../GooglePlaceInput/GooglePlacesInput";
import TripSummaries from "../TripSummaries/TripSummaries";
import {
  Wrapper,
  JourneyInput,
  Heading,
  FormWrapper,
  InputWrapper,
  StartEnd,
  VerticalLine,
  Leaving,
  Button,
} from "./TripPlanner.styled";
import { useJourneyContext } from "../../contexts/journey";
import { useSuggestionsContext } from "../../contexts/suggestions";
import { Location } from "../../../types";

const useSearchParams = () => new URLSearchParams(useLocation().search);

// TODO: type should be Coordinate
const fetchSuggestion = async ({
  from,
  to,
}: {
  from: Location;
  to: Location;
}) => {
  const response = await fetch(
    `${import.meta.env.VITE_TFL_API_URL}/journey/journeyresults/${from
      .coordinates?.latitude},${from.coordinates?.longitude}/to/${to.coordinates
      ?.latitude},${to.coordinates?.longitude}?app_id=1b83c22c&app_key=${
      import.meta.env.VITE_TFL_APP_KEY
    }`,
  );

  return response.json();
};

const TripPlanner = () => {
  const { from, setFrom, to, setTo } = useJourneyContext();
  const { results, reset: resetSuggestions } = useSuggestionsContext();
  const navigate = useNavigate();
  const searchParams = useSearchParams();

  const { data, isLoading } = useQuery({
    queryKey: ["journey", { from, to }],
    queryFn: () => fetchSuggestion({ from: from!, to: to! }),
    enabled: !!from && !!to,
  });

  console.log({ data });

  useEffect(() => {
    if (
      from?.address === "" &&
      to?.address === "" &&
      searchParams.get("toCoordinates") &&
      searchParams.get("fromCoordinates")
    ) {
      const queryParamsFromCoords = {
        lat: parseFloat(
          searchParams.get("fromCoordinates")?.split(",")[0] as string,
        ),
        lon: parseFloat(
          searchParams.get("fromCoordinates")?.split(",")[1] as string,
        ),
      };
      const queryParamsToCoords = {
        lat: parseFloat(
          searchParams.get("toCoordinates")?.split(",")[0] as string,
        ),
        lon: parseFloat(
          searchParams.get("toCoordinates")?.split(",")[1] as string,
        ),
      };
      const queryParamsFromAddress = searchParams.get("fromAddress");
      const queryParamsToAddress = searchParams.get("toAddress");

      // dispatch(fetchSuggestion(queryParamsFromCoords, queryParamsToCoords));
      setFrom({
        address: queryParamsFromAddress as string,
        coordinates: {
          latitude: queryParamsFromCoords.lat,
          longitude: queryParamsFromCoords.lon,
        },
      });
      setTo({
        address: queryParamsToAddress as string,
        coordinates: {
          latitude: queryParamsToCoords.lat,
          longitude: queryParamsToCoords.lon,
        },
      });
    }
  }, [searchParams, from?.address, to?.address]);

  useEffect(() => {
    return () => {
      resetSuggestions();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const handleSubmit = () => {
    if (!from || !to) return;

    navigate(
      `/trip-planner?fromCoordinates=${from?.coordinates?.latitude},${from?.coordinates?.longitude}&toCoordinates=${to?.coordinates?.latitude},${to?.coordinates?.longitude}&fromAddress=${from?.address}&toAddress=${to?.address}`,
    );

    // dispatch(fetchSuggestion(fromCoordinates, toCoordinates));
    fetchSuggestion({ from, to });
  };

  return (
    <Wrapper>
      <JourneyInput>
        <Heading>Let&apos;s take a trip!</Heading>
        <FormWrapper>
          <StartEnd>
            <FontAwesomeIcon
              icon={["fad", "circle"]}
              color="var(--color-green)"
              size="2x"
            />
            <VerticalLine />
            <FontAwesomeIcon
              icon={["fad", "circle"]}
              color="var(--color-background)"
              size="2x"
            />
          </StartEnd>
          <form>
            <InputWrapper>
              {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
              <label htmlFor="fromAddress">From</label>{" "}
              <GooglePlacesInput
                inputId="fromAddress"
                address={from?.address}
                placeholder="Where are you coming from?"
              />
            </InputWrapper>
            <hr />
            <InputWrapper>
              {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
              <label htmlFor="toAddress">To</label>
              <GooglePlacesInput
                inputId="toAddress"
                address={to?.address}
                placeholder="Where are you going to?"
              />
            </InputWrapper>
          </form>
        </FormWrapper>
        <Leaving>
          Leaving now
          <FontAwesomeIcon
            icon={["fad", "chevron-right"]}
            color="var(--color-background)"
            size="2x"
          />
        </Leaving>
        <Button onClick={handleSubmit}>
          {`Search${isLoading ? "ing" : ""}`}
          {isLoading && (
            <FontAwesomeIcon
              icon={["fad", "spinner-third"]}
              color="var(--color-white)"
              size="lg"
            />
          )}
        </Button>
      </JourneyInput>
      {
        // TODO: Disambiguations cause results to be null. Add a way to get exact location instead
        results && results.length ? <TripSummaries journeys={results} /> : null
      }
    </Wrapper>
  );
};

export default TripPlanner;
