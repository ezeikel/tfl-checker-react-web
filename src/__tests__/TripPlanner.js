import React from 'react';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import TripPlanner from '../components/TripPlanner';

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
  useLocation: () => ({
    push: jest.fn(),
  }),
}));

const middlewares = [thunk]; // add your middlewares like `redux-thunk`
const mockStore = configureStore(middlewares);

const setupGoogleMock = () => {
  // Mock Google Maps JavaScript API
  const google = {
    maps: {
      places: {
        AutocompleteService: class{},
        PlacesServiceStatus: {
          INVALID_REQUEST: 'INVALID_REQUEST',
          NOT_FOUND: 'NOT_FOUND',
          OK: 'OK',
          OVER_QUERY_LIMIT: 'OVER_QUERY_LIMIT',
          REQUEST_DENIED: 'REQUEST_DENIED',
          UNKNOWN_ERROR: 'UNKNOWN_ERROR',
          ZERO_RESULTS: 'ZERO_RESULTS',
        },
      },
      Geocoder: () => {},
      GeocoderStatus: {
        ERROR: 'ERROR',
        INVALID_REQUEST: 'INVALID_REQUEST',
        OK: 'OK',
        OVER_QUERY_LIMIT: 'OVER_QUERY_LIMIT',
        REQUEST_DENIED: 'REQUEST_DENIED',
        UNKNOWN_ERROR: 'UNKNOWN_ERROR',
        ZERO_RESULTS: 'ZERO_RESULTS',
      },
    },
  };
  global.window.google = google;
};

beforeAll(() => {
  setupGoogleMock();
});

let mockProps;

beforeEach(() => {
  mockProps = {
    onFetchSuggestions: jest.fn(),
    onSetFromCoords: jest.fn(),
    onSetToCoords: jest.fn(),
    onSetFromAddress: jest.fn(),
    onSetToAddress: jest.fn(),
    onClearSuggestions: jest.fn()
  };
});

test('renders without crashing', () => {
  const store = mockStore({
    journey: {
      from: {
        coordinates: {},
        address: ""
      },
      to: {
        coordinates: {},
        address: ""
      },
    },
    suggestion: {
      results: [],
      loading: false
    }
  });

  const { container } = render(
    <TripPlanner {...mockProps}  store={store} /> // eslint-disable-line react/jsx-props-no-spreading
  );
  expect(container).toMatchSnapshot();
});
