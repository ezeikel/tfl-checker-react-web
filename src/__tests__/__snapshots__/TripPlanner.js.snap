// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <div
    class="sc-fznyAO jRcAa-d"
  >
    <section
      class="sc-fznKkj eKRcEp"
    >
      <span
        class="sc-fznZeY dCAZEX"
      >
        Let's take a trip!
      </span>
      <div
        class="sc-fzokOt hJESQm"
      >
        <span
          class="sc-fzqNJr foHdED"
        >
          <svg
            class="svg-inline--fa fad-circle fa-2x"
            color="var(--color-green)"
          />
          <span
            class="sc-fzoyAV gSsZGr"
          />
          <svg
            class="svg-inline--fa fad-circle fa-2x"
            color="var(--color-background)"
          />
        </span>
        <form>
          <div
            class="sc-fzqBZW iDwreF"
          >
            <label
              for="fromAddress"
            >
              From
            </label>
             
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              autocomplete="off"
              class="sc-AxjAm hhXapv"
              placeholder="Where are you coming from?"
              role="combobox"
              type="text"
              value=""
            />
            <div
              class="sc-AxiKw jAmDvQ"
            >
              <div
                class="sc-AxgMl jGFEmy"
              >
                <div
                  class="sc-AxhUy ixiydF"
                >
                  Places by 
                  <strong>
                    Google
                  </strong>
                </div>
              </div>
            </div>
          </div>
          <hr />
          <div
            class="sc-fzqBZW iDwreF"
          >
            <label
              for="toAddress"
            >
              To
            </label>
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              autocomplete="off"
              class="sc-AxjAm hhXapv"
              placeholder="Where are you going to?"
              role="combobox"
              type="text"
              value=""
            />
            <div
              class="sc-AxiKw jAmDvQ"
            >
              <div
                class="sc-AxgMl jGFEmy"
              >
                <div
                  class="sc-AxhUy ixiydF"
                >
                  Places by 
                  <strong>
                    Google
                  </strong>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
      <span
        class="sc-fzoLag enSKjm"
      >
        Leaving now
        <svg
          class="svg-inline--fa fad-chevron-right fa-2x"
          color="var(--color-background)"
        />
      </span>
      <button
        class="sc-fzoXzr bWBhHs"
      >
        Search
      </button>
    </section>
  </div>
</div>
`;
